[SPARK-49653][SQL] Single join for correlated scalar subqueries

agubichev opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Single join is a left outer join that checks that there is at most 1 build row for every probe row.

This PR adds single join implementation to support correlated scalar subqueries where the optimizer can't guarantee that 1 row is coming from them, e.g.:
select *, (select t1.x from t1 where t1.y >= t_outer.y) from t_outer.
-- this subquery is going to be rewritten as a single join that makes sure there is at most 1 matching build row for every probe row. It will issue a spark runtime error otherwise.

Design doc: https://docs.google.com/document/d/1NTsvtBTB9XvvyRvH62QzWIZuw4hXktALUG1fBP7ha1Q/edit

The optimizer introduces a single join in cases that were previously returning incorrect results (or were unsupported).
Only hash-based implementation is supported, the optimizer makes sure we don't plan a single join as a sort-merge join.


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Expands our subquery coverage.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, previously unsupported scalar subqueries should now work.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Unit tests for the single join operator. Query tests for the subqueries.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No.

agubichev committed:
init

agubichev committed:
comment

agubichev committed:
add test

agubichev committed:
tests

agubichev committed:
done

agubichev committed:
commit

agubichev commented:
@andylam-db @sigmod @cloud-fan

cloud-fan commented:
Do all the existing optimizer rules work fine with this single join? I understand that we need to implement the single-match check in all the physical join nodes, but semantic wise, is there anything we need to take care?

agubichev commented:
> Do all the existing optimizer rules work fine with this single join? I understand that we need to implement the single-match check in all the physical join nodes, but semantic wise, is there anything we need to take care?

@cloud-fan 

I've traced all the usages of LeftOuter in the catalyst rules (see the full list below). In general, the rules act on the basis of "allow-list", so if the join type is not explicitly matched by the rule, it is not applied. As LeftOuter is a "close relative" to LeftSingle (in fact, at HEAD we are using LeftOuter in place of LeftSingle), it is enough to check the rules that already reference LeftOuter explicitly. Since LeftOuter joins are already super restrictive as to what kind of optimizations can be applied to them (and frequently LeftOuter joins restrict optimizations in the plan nodes around them too), I am not aware of many jointype-agnostic rules. The ones that I do know of, like ReplaceNullWithFalseInPredicate, apply to both LeftOuter and LeftSingle without change. 

These rules have been updated for LeftSingle join:

- EliminateOuterJoin -- should not apply to LeftSingle, updated
- PushPredicateThroughJoin -- not all cases should apply to LeftSingle, updated
- FoldablePropagation

The following rules are only matching LeftOuter join for now, therefore skipping LeftSingle join unchanged. Semantics-wise, it is ok to skip every single one of these rules for the LeftSingle join. Further analysis is needed on whether we can/should enable them for LeftSingle joins:

- InferFiltersFromConstraints
- LimitPushDown
- PropagateEmptyRelation
- PushLeftSemiLeftAntiThroughJoin
- PushExtraPredicateThroughJoin

There are couple of rules that apply to LeftOuter, but do not make sense to LeftSingle. In both cases LeftOuter is explicitly matched, so they will skip LeftSingle as they should:

- CheckCartesianProducts
- RewriteAsOfJoin

agubichev committed:
typo

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -1046,10 +1049,12 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
-                  "MUST_AGGREGATE_CORRELATED_SCALAR_SUBQUERY",
-                messageParameters = Map.empty)
+              if (!SQLConf.get.getConf(SQLConf.SCALAR_SUBQUERY_USE_SINGLE_JOIN)) {
+                expr.failAnalysis(
```
just for my curiosity, why there are two places checking scalar subqueries?

agubichev replied:
The second check (in lines 1060-1080 of the current file) checks the places where the scalar subquery is allowed (e.g., it can occur in the project or filter but not in the join predicate)

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:
```diff
@@ -135,8 +138,15 @@ case class BroadcastNestedLoopJoinExec(
-  private def outerJoin(relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+  private def outerJoin(
+      relation: Broadcast[Array[InternalRow]],
+      checkMatches: Int => Int): RDD[InternalRow] = {
```
shouldn't it be `Int => Unit`?

agubichev replied:
it also increments the number of matches and returns the updated count.
Otherwise we would need to increment it in the body of the join and do it for LeftSingle AND LeftOuter joins. This way we avoid incrementing the counter for LeftOuter join where it is not needed.

agubichev replied:
removed the checkMatches in favor of your next suggestion.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:
```diff
@@ -162,12 +174,14 @@ case class BroadcastNestedLoopJoinExec(
             nextIndex += 1
             if (boundCondition(resultRow)) {
               foundMatch = true
+              matches = checkMatches(matches)
```
Seems we only need a single `singleJoin` boolean flag?
```
if (boundCondition(resultRow)) {
  if (foundMatch && singleJoin) throw ...
  foundMatch = true
}
```

agubichev replied:
neat!

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-group-by.sql:
```diff
@@ -22,16 +22,27 @@ select *, (select count(*) from y where x1 = y1 and cast(y2 as double) = x1 + 1
+set spark.sql.optimizer.scalarSubqueryUseSingleJoin = false;
 select * from x where (select count(*) from y where y1 > x1 group by y1) = 1;
 reset spark.sql.legacy.scalarSubqueryAllowGroupByNonEqualityCorrelatedPredicate;
+reset spark.sql.optimizer.scalarSubqueryUseSingleJoin;
```
it's actually not needed as each golden file is tested with a fresh SparkSession

agubichev replied:
removed

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/joins/SingleJoinSuite.scala:
```diff
@@ -0,0 +1,213 @@
+
+
+  private def testSingleJoin(
+                              testName: String,
```
nit: 4 spaces indentation

agubichev replied:
done

agubichev commented on sql/connect/common/src/main/protobuf/spark/connect/relations.proto:
```diff
@@ -258,6 +258,7 @@ message Join {
     JOIN_TYPE_LEFT_ANTI = 5;
     JOIN_TYPE_LEFT_SEMI = 6;
     JOIN_TYPE_CROSS = 7;
+    JOIN_TYPE_LEFT_SINGLE = 8;
```
removed

agubichev commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -269,8 +269,13 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         }
 
+        def canMerge(joinType: JoinType): Boolean = joinType match {
+          case LeftSingle => false
```
in principle, it can do it, but I did not implement it.
as we only need single joins for correlated scalar subqueries with non-equal predicate, the SMJ are unlikely to be useful there anyways.
(for equi-predicates like 
'col = outer(col) .. group by col', we reason that the query returns at most 1 row and don't need the single join).

agubichev commented on sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-group-by.sql:
```diff
@@ -22,16 +22,27 @@ select *, (select count(*) from y where x1 = y1 and cast(y2 as double) = x1 + 1
 select * from x where (select count(*) from y where y1 > x1 group by y1) = 1;
 select *, (select count(*) from y where y1 + y2 = x1 group by y1) from x;
 select *, (select count(*) from y where x1 = y1 and y2 + 10 = x1 + 1 group by y2) from x;
+reset spark.sql.optimizer.scalarSubqueryUseSingleJoin;
```
removed

agubichev commented on sql/core/src/test/scala/org/apache/spark/sql/execution/joins/SingleJoinSuite.scala:
```diff
@@ -0,0 +1,213 @@
+
+    testWithWholeStageCodegenOnAndOff(s"$testName using BroadcastHashJoin") { _ =>
+      extractJoinParts().foreach { case (_, leftKeys, rightKeys, boundCondition, _, _, _, _) =>
+        withSQLConf(SQLConf.SHUFFLE_PARTITIONS.key -> "1") {
```
no reason, just copied it from similar test. Removed it now

agubichev committed:
done

agubichev commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -456,6 +456,33 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
   }
 
+   // Returns true if 'query' is guaranteed to return at most 1 row.
+   private def guaranteedToReturnOneRow(query: LogicalPlan): Boolean = {
```
done, but the amount of savings is not spectacular, because in CheckAnalysis there is also a "legacy" path (that results in incorrect results) where the check is significantly weaker.

Once the single join is rolled out, we will just remove all extra checks from CheckAnalysis.

agubichev committed:
cleanup

cloud-fan commented:
thanks, merging to master!
